libs/core/algorithms/include/hpx/parallel/datapar/replace.hpp

```diff
@@ -0,0 +1,306 @@
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/config.hpp>
+
+#if defined(HPX_HAVE_DATAPAR)
+#include <hpx/execution/traits/is_execution_policy.hpp>
+#include <hpx/execution/traits/vector_pack_conditionals.hpp>
+#include <hpx/executors/datapar/execution_policy.hpp>
+#include <hpx/functional/tag_invoke.hpp>
+#include <hpx/parallel/algorithms/detail/distance.hpp>
+#include <hpx/parallel/algorithms/detail/replace.hpp>
+#include <hpx/parallel/datapar/handle_local_exceptions.hpp>
+#include <hpx/parallel/datapar/loop.hpp>
+#include <hpx/parallel/util/result_types.hpp>
+#include <hpx/parallel/util/zip_iterator.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 { namespace detail {
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename ExPolicy>
+    struct datapar_replace
+    {
+        template <typename InIter, typename T1, typename T2, typename Proj>
+        HPX_HOST_DEVICE HPX_FORCEINLINE static auto call(ExPolicy&& policy,
+            InIter first, InIter last, T1 const& old_value, T2 const& new_value,
+            Proj&& proj)
+        {
+            if constexpr (hpx::is_sequenced_execution_policy_v<ExPolicy>)
+            {
+                return util::loop_ind(HPX_FORWARD(ExPolicy, policy), first,
+                    last, [old_value, new_value, &proj](auto& v) {
+                        using var_type = std::decay_t<decltype(v)>;
+                        traits::mask_assign(
+                            HPX_INVOKE(proj, v) == var_type(old_value), v,
+                            var_type(new_value));
+                    });
+            }
+            else
+            {
+                return for_each_n<InIter>().call(
+                    HPX_FORWARD(ExPolicy, policy), first,
+                    std::distance(first, last),
+                    [old_value, new_value, proj = HPX_FORWARD(Proj, proj)](
+                        auto& v) -> void {
+                        traits::mask_assign(
+                            HPX_INVOKE(proj, v) == var_type(old_value), v,
+                            var_type(new_value));
+                    },
+                    util::projection_identity());
+            }
+        }
+    };
+
+    template <typename ExPolicy, typename InIter, typename T1, typename T2,
+        typename Proj,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_vectorpack_execution_policy<ExPolicy>::value)>
+    HPX_HOST_DEVICE HPX_FORCEINLINE auto tag_invoke(
+        sequential_replace_t<ExPolicy>, ExPolicy&& policy, InIter first,
+        InIter last, T1 const& old_value, T2 const& new_value, Proj&& proj)
+    {
+        if constexpr (hpx::parallel::util::detail::iterator_datapar_compatible<
```

Comment:
Could you add this to the `HPX_CONCEPT_REQUIRES_` above, thus avoiding duplicating the `else` case below? This might also require adding the `to_non_simd` to the non-datapar case, though.

Reply:
Let's leave this PR as is and get back to the mentioned issue later.